File: crates/legion-loop/src/context.rs
//! Context window builder: converts stored `TurnEnvelope`s into rs-ai `Message`s.

use legion_core::types::{TurnEnvelope, TurnEventKind};
use rs_ai::types::{ContentBlock, Message, Role};

/// Build an rs-ai message list from the recent turn history.
///
/// Only `UserMessage`, `AssistantMessage`, and `ToolResult` turns are included;
/// write-ahead intents and lifecycle events are skipped.
pub fn build_messages(envelopes: &[TurnEnvelope]) -> Vec<Message> {
    let mut messages = Vec::new();

    for env in envelopes {
        match &env.event.kind {
            TurnEventKind::UserMessage => {
                let text = extract_text_content(&env.event.payload);
                messages.push(user_message(text));
            }
            TurnEventKind::AssistantMessage => {
                let text = extract_text_content(&env.event.payload);
                messages.push(assistant_message(text));
            }
            TurnEventKind::ToolResult { call_id } => {
                let result = env.event.payload
                    .as_ref()
                    .map(|v| v.to_string())
                    .unwrap_or_default();
                messages.push(tool_result_message(
                    call_id,
                    env.event.model.as_deref().unwrap_or(""),
                    result,
                    false,
                ));
            }
            // Skip intents and lifecycle events — not LLM-visible
            _ => {}
        }
    }
    messages
}

fn extract_text_content(payload: &Option<serde_json::Value>) -> String {
    match payload {
        Some(v) => {
            if let Some(s) = v.get("content").and_then(|c| c.as_str()) {
                s.to_string()
            } else {
                v.to_string()
            }
        }
        None => String::new(),
    }
}

fn blank_message(role: Role, text: String) -> Message {
    Message {
        role,
        content: vec![ContentBlock::Text {
            text,
            text_signature: None,
        }],
        timestamp:        0,
        api:              None,
        provider:         None,
        model:            None,
        response_id:      None,
        response_model:   None,
        diagnostics:      Vec::new(),
        usage:            None,
        stop_reason:      None,
        deferred:         None,
        error_message:    None,
        raw_stop_reason:  None,
        end_turn:         None,
        tool_call_id:     None,
        tool_name:        None,
        is_error:         false,
        details:          None,
        added_tool_names: Vec::new(),
    }
}

pub fn user_message(text: String) -> Message {
    blank_message(Role::User, text)
}

pub fn assistant_message(text: String) -> Message {
    blank_message(Role::Assistant, text)
}

pub fn tool_result_message(call_id: &str, tool_name: &str, result: String, is_error: bool) -> Message {
    Message {
        role: Role::ToolResult,
        content: vec![ContentBlock::Text {
            text: result,
            text_signature: None,
        }],
        timestamp:        0,
        api:              None,
        provider:         None,
        model:            None,
        response_id:      None,
        response_model:   None,
        diagnostics:      Vec::new(),
        usage:            None,
        stop_reason:      None,
        deferred:         None,
        error_message:    None,
        raw_stop_reason:  None,
        end_turn:         None,
        tool_call_id:     Some(call_id.to_string()),
        tool_name:        Some(tool_name.to_string()),
        is_error,
        details:          None,
        added_tool_names: Vec::new(),
    }
}

File: internal/agent/context.go
package agent

import (
	"encoding/json"

	goai "github.com/rcarmo/go-ai"
	"github.com/rcarmo/legion/internal/core"
)

func buildContext(log []core.TurnEnvelope, config core.RunConfig) *goai.Context {
	ctx := &goai.Context{}
	if config.SystemPrompt != nil {
		ctx.SystemPrompt = *config.SystemPrompt
	}
	for i, e := range log {
		switch e.Event.Kind.Kind {
		case "user_message":
			var p struct {
				Content string `json:"content"`
			}
			_ = json.Unmarshal(e.Event.Payload, &p)
			ctx.Messages = append(ctx.Messages, goai.UserMessage(p.Content))
		case "assistant_message":
			var p struct {
				Message *goai.Message `json:"message"`
				Content string        `json:"content"`
			}
			_ = json.Unmarshal(e.Event.Payload, &p)
			if p.Message != nil {
				ctx.Messages = append(ctx.Messages, *p.Message)
			} else {
				ctx.Messages = append(ctx.Messages, goai.Message{Role: goai.RoleAssistant, Content: []goai.ContentBlock{{Type: "text", Text: p.Content}}})
			}
		case "tool_result":
			name := ""
			for j := i - 1; j >= 0; j-- {
				k := log[j].Event.Kind
				if k.Kind == "tool_call_intent" && k.CallID == e.Event.Kind.CallID {
					name = k.ToolName
					break
				}
			}
			var value any
			_ = json.Unmarshal(e.Event.Payload, &value)
			b, _ := json.Marshal(value)
			goai.AppendToolResult(ctx, e.Event.Kind.CallID, name, string(b), isError(value))
		}
	}
	return ctx
}
func isError(v any) bool {
	m, ok := v.(map[string]any)
	if !ok {
		return false
	}
	_, ok = m["error"]
	return ok
}
